---
title: "create_seurat_objects_from_kallisto_output" # from kallisto output
output: html_document
---

# this workflow reads kallisto/bustools matrix files, removes empty droplets & creates seurat objects, which will be saved under kb_data/out/"index"
# based on seurat 4.0, dropletutils 1.9.0, busparse 1.3.0

# file structure: ~/kb_data/fastq/    --> place fastq files here; no need to unpack; '.fastq.gz'
#                 ~/kb_data/ref-seqs/ --> reference transcriptome data for kallisto index (.fa.gz & .gtf.gz)
#                 ~/kb_data/index/    --> contains the created index files; T-DBG
#                 ~/kb_data/out/     --> output directory

```{r message=FALSE, warning=FALSE}
#load packages
library(BUSpaRse)
library(tidyverse)
library(stringr)
library(ggplot2)
library(dplyr)
library(Seurat)
library(DropletUtils)
library(Matrix)

# SETUP

expers <- c("sample_id"
            )

index <- c("reference"
           )

EXAMPLES
# expers <- c("dr_RGC_larva_s1", "dr_RGC_adult_s17", "dr_pineal_s1", "dr_pineal_s2",
#             "hs_pbmc_4k_v2", "hs_pbmc_8k_v2", "hs_pbmc_1k_v3", "hs_pbmc_10k_v3"
#             )
# 
# index <- c('D_rerio.GRCz11.101', 'D_rerio.GRCz11.101', 'D_rerio.GRCz11.101', 'D_rerio.GRCz11.101',
#            "H_sapiens.GRCh38.100", "H_sapiens.GRCh38.100", "H_sapiens.GRCh38.100", "H_sapiens.GRCh38.100"
#            )

project <- paste(expers, "_kb", sep="")

#setup working directory
knitr::opts_knit$set(root.dir = '~/kb_data/')
```


# Remove empty droplets and create seurat objects
```{r}
i <- 0
for (e in expers) {
  i <- i + 1
  print(paste0("Working on: ", e))
  res_mat <- read_count_output(str_c("~/kb_data/out/", index[i], "/", e, "/counts_unfiltered/"), name = "cells_x_genes", tcc = FALSE)
  dim(res_mat)
  tot_counts <- Matrix::colSums(res_mat)
  summary(tot_counts)
  
# Compute barcode ranks for both samples & remove empty droplets
  bc_rank <- barcodeRanks(res_mat, lower = 500)

knee_plot <- function(bc_rank) {
  knee_plt <- tibble(rank = bc_rank[["rank"]],
                     total = bc_rank[["total"]]) %>% 
    distinct() %>% 
    dplyr::filter(total > 0)
  annot <- tibble(inflection = metadata(bc_rank)[["inflection"]],
                  rank_cutoff = max(bc_rank$rank[bc_rank$total > metadata(bc_rank)[["inflection"]]]))
  p <- ggplot(knee_plt, aes(total, rank)) +
    geom_line() +
    geom_hline(aes(yintercept = rank_cutoff), data = annot, linetype = 2) +
    geom_vline(aes(xintercept = inflection), data = annot, linetype = 2) +
    scale_x_log10() +
    scale_y_log10() +
    annotation_logticks() +
    labs(y = "Rank", x = "Total UMIs")
  ggsave(str_c("~/kb_data/out/", index[i], "/", e, "/", e, "_knee_plot_kb.png"), last_plot(), device="png", width = 9, height = 6)
  return(p)
}
options(repr.plot.width=9, repr.plot.height=6)
knee_plot(bc_rank)

  # Remove genes that are not detected and empty droplets
  res_mat <- res_mat[, tot_counts > metadata(bc_rank)$inflection]
  dim(res_mat) # use print() to show output in a for loop
  
  seu <- CreateSeuratObject(res_mat, min.cells = 3, min.feature = 200, project = project[i])
  saveRDS(seu, file = str_c("~/kb_data/out/", index[i], "/", e, "_kb.rds"))  # seu.rds is saved
  print(seu)
}
```